viewing_party/party.py

```diff
@@ -1,23 +1,252 @@
+    Find the sum of the integers in value for "rating" in the movie dict
+    If the input is an empty list, return 0.0 for average
+    Return the sum of movie ratings divided by the number of movies in input user_data dict
+    """
+
+    sum_rating = 0
+    
+    movie_list = user_data["watched"]
+
+    for movie in movie_list:
+        sum_rating += movie["rating"]
+    if len(movie_list) == 0:
+        return 0.0
+    else:
+        return sum_rating / len(movie_list)
+
+
+def get_most_watched_genre(user_data):
+    """
+    Create variable that is a list
+    Set variable to user's watched list
+    Loop through movies in user's watched list
+    Add movie genre to genre list
+    If genre list has 0 genres, then return None
+    Multimode returns most commonly occurring element in list
+
+    """
+    genre_list = []
+
+    movie_list = user_data["watched"]
+
+    for movie in movie_list:
+        genre_list.append(movie["genre"])
+
+    if len(genre_list) == 0:
+        return None
+    else:
+        return(multimode(genre_list)[0])
+
+
+# # My super long implementation w/o using multimode
+# def get_most_watched_genre(user_data):
+#     # Get the watched list
+#     watched_list = user_data["watched"]
+
+#     # Check if the watched list is empty
+#     if len(watched_list) == 0:
+#         return None
+
+#     # Create a dictionary to store the count of each genre
+#     genre_count = {}
+
+#     # Iterate through the watched list
+#     for movie in watched_list:
+#         genre = movie["genre"]
+
+#     # Check if the genre is already in the dictionary
+#     if genre in genre_count:
+#         # Increment the count
+#         genre_count[genre] += 1
+#     else:
+#         # Initialize the count
+#         genre_count[genre] = 1
+
+#     # Get the most frequent genre
+#     most_frequent_genre = None
+#     most_frequent_count = 0
+
+#     for genre, count in genre_count.items():
+#         if count > most_frequent_count:
+#             most_frequent_count = count
+
+#     return most_frequent_count
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    """ 
+    Get the list of watched movies from user
+    Create variable of friends watched movies
+    Set variable to list of unique movies
+    Loop through friends 
+    Adding friends watched movies to list friends_watched
+    Iterate through movies of user's watched list
+    Conditional that movie is not in friends watched and not in unique movies lists
+    Add that movie to unique movies list
+    """
+
+    user_watched = user_data["watched"]
+    friends_watched = []
+    unique_movies = []
+
+    for friend in user_data["friends"]:
+        friends_watched += friend["watched"]
+
+    for movie in user_watched:
+        if movie not in friends_watched and movie not in unique_movies:
+            unique_movies.append(movie)

```

Comment:
Consider what the big-O runtime of this loop might be. It turns out that using `in` on a list actually hides a loop, so as the combined friends list gets longer, the length of time to see whether the next movie is already in the list also takes longer. We could use a helper set of movie titles to help us do this check more efficiently. Each time we add a movie to the combined list, we could add its title to a set. Then, when deciding whether or not to add the next movie, we could check whether that movie's title is in the title set already. `in` on sets is much more efficient than on lists.

Reply:
It'd be a good idea to think about the time complexity of the other methods that use `in`, too.